Add new user gallery feature to show user video gallery list

javierLiarte opened the pull request:
Merge firs #38 into this branch!

- Add main controller, user, gallery and page services, video list directive and user gallery controller to index.html
- Add title to index.html
- Add userGallery state to app.js
- Added new strings for user gallery to lang files
- Modify gallery controller to use new gallery service
- Modify gallery view html to use new video-list directive
- Add MainCtlr to inject Page in scope
- Create Page service to set window title
- Create UserService to get user details from a user id
- Create new GalleryService for getting paginated video lists
- Add tests for GalleryService
- Create new user-gallery page to show user video list
- Create new VideoList directive for showing video gallery lists
- Add new pathspecs to sourceJs and sourceTest in gulpfile to solve watch issues

javierLiarte committed:
Add new user gallery feature to show user video gallery list

- Add main controller, user, gallery and page services, video list directive and user gallery controller to index.html
- Add title to index.html
- Add userGallery state to app.js
- Added new strings for user gallery to lang files
- Modify gallery controller to use new gallery service
- Modify gallery view html to use new video-list directive
- Add MainCtlr to inject Page in scope
- Create Page service to set window title
- Create UserService to get user details from a user id
- Create new GalleryService for getting paginated video lists
- Add tests for GalleryService
- Create new user-gallery page to show user video list
- Create new VideoList directive for showing video gallery lists
- Add new pathspecs to sourceJs and sourceTest in gulpfile to solve watch issues

javierLiarte commented:
There's still some work pending on styles for new views and directive.

Please be gently as this is my first story in angularJS :P

CodingCarlos commented on services/gallery.js:
```diff
@@ -0,0 +1,55 @@
+class GalleryService {
```
We shouldn't use ES6 in client since it's not supported by all browsers. That's the same reason we're using sass polyfills.

javierLiarte replied:
With that in mind I've checked compatibility issues of EC6 first. Since it's been almost three years now from its release in june '15, it's widely supported now as can be seen on https://kangax.github.io/compat-table/es6/ and thought it was safe to use.

(If we were still worried about compatibility issues after seeing that compatibility table, it's also possible to use transpilers like babel loader with webpack.)

I'm think it's a pitty to loose all the advantages and progresses that ES6 provides for this concerns.

CodingCarlos replied:
If compatibility is not concerning, so we shall see if we need sass compatibility mixins. As I see, the issues are in IE11, Android browser, Samsung browser, firefox < 45 and chrome <49. This means about a 17% ([source](https://caniuse.com/#search=es6%20classes).)

In any case, sytactic suggar `class` is more a disadvantage than a progress, since there aren't real oop classes, but just the classic prototypes js used to have. So, if using the same concept, why calling it in a different and confuse way? Just my view on this.

javierLiarte replied:
For me it simplify syntax and improves readability of code. As can I use states: ES6 classes are syntactical sugar to provide a much simpler and clearer syntax to create objects and deal with inheritance. ;)

Anyway, as I was saying before, we can use babel transpiler to cover non supported browsers also.

CodingCarlos commented on services/gallery.js:
```diff
@@ -0,0 +1,55 @@
+function galleryServiceGenerator(api, $httpParamSerializer) {
+	return {
+		getInstance: function (tagFilter, userId) {
+			return new GalleryService(api, $httpParamSerializer, tagFilter, userId);
```
Services are singletons, but we're using a singleton to instanciate new modules, which is not the idea of services. To do that, angular have other patterns as the own directives, or even controllers. In this case, I think we coud find a better way to do this.

javierLiarte replied:
I understand know why it would be hard to find examples of factories in angular. So factory in angular is not intended to build factories?

Current "service" implementation makes sense to me with current services developed, such as api or user services, that makes sense to be singletons. What I tried to make here is another thing on top of a service, the gallery factory, that, using api service, will provide video lists paginated support, and can then generalize existing code for easier reusing (such as in exisiting gallery: one instance for handling featured videos, another for handling non featured separately; think also about youtube, that holds different separate video lists) and better maintainability.

So I will need help choosing the best angular construct to implement this approach

CodingCarlos commented on services/gallery.js:
```diff
@@ -0,0 +1,55 @@
+}
+
+angular.module('app')
+	.service('galleryServiceFactory', ['api', '$httpParamSerializer', galleryServiceGenerator]);
```
We're naming services just by the function it have, without including the `service` or `factory` word in the name. So,  `galleryServiceFactory` might only be `gallery`. I guess you used that name, becouse it's like a generator pattern, but, as I've said in the previous comment, we shall refactor this service in somehing different.

javierLiarte replied:
That's is, because this approach is a gallery factory, and thus, not a service, as I understood

CodingCarlos commented on services/user.js:
```diff
@@ -0,0 +1,23 @@
+(function () {
+	class UserService {
```
The same ES6 and browser compatibility comment I've already added ;)

CodingCarlos commented on services/user.js:
```diff
@@ -0,0 +1,23 @@
+		}
+
+		getUser(userId) {
+			return new Promise((resolve, reject) => {
```
Returning promises is not consistent with any service we already have, since we're actually returning callbacks in every service

CodingCarlos reviewed (changes requested):
In summary:

 - Services are thought for singleton patterns. If we don't want to use a singleton, we shall not use a service, but someting like controllers or directive's scopes.
 - ES6 is dangerous to use in browser, since it's not fully supported. If we're using poifills in SASS for things like flexbox, we shall follow the same compatibility strategy in JS.
 - We're actually using callbacks to return data from services, instead of promises. We shall be consistent in one or other way.

javierLiarte committed:
Change MainController and page service to not use scope and to follow name conventions

- Update index.html to use new page service and main controller as main.page.title
- Change page service angular name to be undercase. Removed setters and getters and use directly page title field
- Update Main controller to stop using $scope for setting page service
- Update user gallery controller to use new page service title field directly
- Fix several linting issues

DevStarlight commented on services/gallery.js:
```diff
@@ -0,0 +1,55 @@
+}
+
+angular.module('app')
+	.service('galleryServiceFactory', ['api', '$httpParamSerializer', galleryServiceGenerator]);
```
Okay taking gallery out of its own page folder and making it global, but remove the local gallery service since it's not used anymore

DevStarlight commented on components/video-list/video-list-non-featured.view.html:
```diff
@@ -0,0 +1,12 @@
+<div infinite-scroll="infinite()" infinite-scroll-distance="1" infinite-scroll-disabled="loading">
+    <section class="{{sectionClass}}">
```
remove some watchers by just adding :: to the name of the variable like this: {{::sectionClass}} since this is gonna be an static value

DevStarlight commented on components/video-list/video-list-non-featured.view.html:
```diff
@@ -0,0 +1,12 @@
+            <video-card ng-repeat="video in datasource"
+                        ng-style="($index + 1) % videosPerRow === 0 ? { 'margin':'0px' } : {}"
+                        video="video"
+                        width="width" height="height"
```
same here... `::width` and `::height`

javierLiarte committed:
Refactor of old gallery factory service to include it into video-list component

- Refactor of video list directive to a component now using its own controller instead of a gallery service factory
- Update gallery and user gallery controller to use new video-list component; remove code for getting videos from api using old gallery service factory
- Update gallery and user gallery view to use new video-list component attributes
- Add css/style.css to git ignore

javierLiarte commented:
Refactored video list component to stop using old gallery factory.

Pending to:
* merge last dev changes into branch
* remove old gallery service
* adapt old service tests to test new class
* fix gallery and user gallery controller tests
* @DevStarlight is currently working on gallery css to remove calculations from page controllers

Also if we want to disable ES6 classes until we discuss it or include babel or similar transpiler, I can make a fast refactor of current classes

DevStarlight committed:
continuation of gallery refactor, this time for design.

DevStarlight commented on components/video-list/video-list.view.html:
```diff
@@ -0,0 +1,12 @@
+<div infinite-scroll="VideoList.infinite()" infinite-scroll-distance="1" infinite-scroll-disabled="VideoList.loading">
+    <section class="{{VideoList.sectionClass}}">
+        <article>
```
We are creating an article in this directive.... that creates multiple articles in case that it is called twice

DevStarlight commented on components/video-list/video-list.view.html:
```diff
@@ -0,0 +1,12 @@
+<div infinite-scroll="VideoList.infinite()" infinite-scroll-distance="1" infinite-scroll-disabled="VideoList.loading">
+    <section class="{{VideoList.sectionClass}}">
+        <article>
+            <video-card ng-repeat="video in VideoList.videos"
```
there is a lot of stuff that could be deleted like ng-style, width, height...

javierLiarte committed:
Fix video list component infinite scrolling disable var

- Fix gallery infinte scroll disable var in video-list controller
- Enable infinite scrolling in gallery view for non-featured video gallery

javierLiarte committed:
Remove old unused services. Fix some tests

- Remove old gallery service and page service imports from index.html
- Remove old gallery service and gallery page service
- Fix user gallery controller test
- Remove ignored file css/style.css
- Move old gallery service tests to video list component folder

javierLiarte committed:
Merge develop into 155463483_user_gallery

javierLiarte committed:
Fix video-list tag-filter definition in gallery view. Recover user gallery and upload controllers import in index.html

javierLiarte committed:
Fix failing video list and register tests

- Fix video list component controller tests
- Fix register failing test

DevStarlight committed:
updated user-gallery

DevStarlight committed:
Merge branch '155463483_user_gallery' of github.com:Videona/platform-frontend into 155463483_user_gallery

# Conflicts:
#	components/topbar/topbar.scss

javierLiarte committed:
Add missing changes fixing register test

javierLiarte committed:
Fix user gallery title when user has not username

DevStarlight commented on pages/user-gallery/user-gallery.controller.js:
```diff
@@ -18,11 +19,12 @@
-				page.title = self.user.username + ' User gallery - ';
-				$translate('USER_GALLERY_USER_VIDEOS', { username: self.user.username })
-					.then(title => self.userGalleryTitle = title);
+				if (typeof self.user !== 'undefined') {
```
Isn't that suppose that before calling the server in userService, we should check the userId so if there's undefined value going there, reject the promise? I don't agree with this @javierLiarte

javierLiarte replied:
ok, I deal with this now :)

DevStarlight committed:
change body height size

javierLiarte committed:
Handle userId param not defined or not number in user gallery controller

javierLiarte committed:
Merge pull request #38 from Videona/fixInfiniteScroll

Fix infinite scroll
